pennylane/devices/default_clifford.py

```diff
@@ -0,0 +1,726 @@
+                        else:
+                            debugger.snapshots[len(debugger.snapshots)] = flat_state
+
+        tableau_simulator.do_circuit(stim_ct)
+
+        global_phase = qml.GlobalPhase(qml.math.sum(op.data[0] for op in global_phase_ops))
+
+        results = []
+        for meas in circuit.measurements:
+            # Analytic case
+            if not circuit.shots:
+                # Computing statevector via tableaus
+                if type(meas) is qml.measurements.StateMP:
+                    res = self._measure_state(tableau_simulator, interface, global_phase)
+
+                # Computing density matrix via tableaus
+                elif type(meas) is qml.measurements.DensityMatrixMP:
+                    res = self._measure_density_matrix(tableau_simulator, interface)
+
+                # Computing purity via tableaus
+                elif type(meas) is qml.measurements.PurityMP:
+                    res = self._measure_purity(tableau_simulator, interface, meas, circuit)
+
+                # Computing expectation values via measurement
+                elif isinstance(meas, ExpectationMP):
+                    res = self._measure_expectation(tableau_simulator, interface, meas, stim)
+
+                # Computing more measurements
+                else:
+                    raise NotImplementedError(
+                        f"default.clifford doesn't support the {type(meas)} measurement at the moment."
+                    )
+
+                results.append(res)
+
+        return tuple(results)
+
+    def _measure_state(self, tableau_sim, interface, global_phase):
+        """Measure the state of the simualtor device"""
+        if self._state == "tableau":
+            tableau = tableau_sim.current_inverse_tableau().inverse()
+            x2x, x2z, z2x, z2z, x_signs, z_signs = tableau.to_numpy()
+            return np.vstack(
+                (
+                    np.hstack((x2x, x2z, x_signs.reshape(-1, 1))),
+                    np.hstack((z2x, z2z, z_signs.reshape(-1, 1))),
+                )
+            ).astype(int)
+
+        return qml.math.array(
+            tableau_sim.state_vector(endian="big"),
+            like=INTERFACE_TO_LIKE[interface],
+        ) * qml.matrix(global_phase)
+
+    @staticmethod
+    def _measure_density_matrix(tableau_sim, interface):
+        """Measure the density matrix from the state of simulator device"""
+        state_vector = qml.math.array(
+            tableau_sim.state_vector(endian="big"),
+            like=INTERFACE_TO_LIKE[interface],
+        )
+        return qml.math.einsum("i, j->ij", state_vector, state_vector)
+
+    @staticmethod
+    def _measure_purity(tableau_sim, interface, meas_op, circuit):
+        """Measure the purity of the state of simulator device"""
+        if circuit.op_wires == meas_op.wires:  # // Trivial
+            purity = qml.math.array(1.0, like=INTERFACE_TO_LIKE[interface])
+        return purity
```

Comment:
If we are measuring all wires, then the purity will always be one since the state will always be pure (in the noiseless setting). The non-trivial case is not straightforward as of now, so haven't added those.

Reply:
```suggestion
        if circuit.op_wires == meas_op.wires:  # // Trivial
            purity = qml.math.array(1.0, like=INTERFACE_TO_LIKE[interface])
        else:
            raise NotImplementedError
        return purity
```
I'd rather be explicit about this branch not being supported than deal with inevitable and less informative `NameError`.

Reply:
We had decided to separate out measurement support into two PRs. I'll add it there.